File: ios/Classes/FlutterAcsCardReaderPlugin.swift
import Flutter
import UIKit

public class FlutterAcsCardReaderPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel!
    private var smartCardReader: SmartCardReader!
    private var isScanning: Bool = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_acs_card_reader", binaryMessenger: registrar.messenger())
        let instance = FlutterAcsCardReaderPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        instance.smartCardReader = SmartCardReader(methodChannel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "connectToDevice":
            guard let arguments = call.arguments as? [String: Any],
                  let driver = arguments["driver"] as? [String: Any],
                  let cardTerminalType = arguments["cardTerminalType"] as? Int,
                  let timeoutSeconds = arguments["timeoutSeconds"] as? Int else {
                result(FlutterError(code: "INVALID_DEVICE", message: "Invalid driver, cardTerminalType, or timeoutSeconds parameter", details: nil))
                return
            }
            connectToDevice(result: result, driver: driver, cardTerminalType: cardTerminalType, timeoutSeconds: timeoutSeconds)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func connectToDevice(result: @escaping FlutterResult, driver: [String: Any], cardTerminalType: Int, timeoutSeconds: Int) {
        guard let card = driver["card"] as? String,
              let name = driver["name"] as? String,
              let firstName = driver["firstName"] as? String,
              let email = driver["email"] as? String,
              let phone = driver["phone"] as? String,
              let agencyID = driver["agencyID"] as? String else {
            result(FlutterError(code: "INVALID_DEVICE", message: "Invalid parameters", details: nil))
            return
        }

        let driver = Driver(card: card, name: name, firstName: firstName, email: email, phone: phone, agencyID: agencyID)
        print("DRIVER \(driver) cardTerminalType \(cardTerminalType) timeoutSeconds \(timeoutSeconds)");
        smartCardReader.connectToDevice(driver: driver, cardTerminalType: cardTerminalType, timeoutSeconds: timeoutSeconds)
    }
}

File: android/src/main/kotlin/com/benjamin/horner/flutter_acs_card_reader/FlutterAcsCardReaderPlugin.kt
package com.benjamin.horner.flutter_acs_card_reader

/// Local
import com.benjamin.horner.flutter_acs_card_reader.SmartCardReader
import com.benjamin.horner.flutter_acs_card_reader.Driver

/// Android
import android.content.Context
import android.app.Activity
import android.util.Log
import androidx.annotation.NonNull

/// Flutter 
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class FlutterAcsCardReaderPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var activity: Activity
    private lateinit var smartCardReader: SmartCardReader
    private var isScanning: Boolean = false

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_acs_card_reader")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        smartCardReader = SmartCardReader(channel)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity;
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "connectToDevice" -> {
                val driver = call.argument<Map<String, Any>>("driver")
                val cardTerminalType = call.argument<Int>("cardTerminalType")
                val timeoutSeconds = call.argument<Int>("timeoutSeconds")
                if (driver != null && cardTerminalType != null && timeoutSeconds != null) {
                    connectToDevice(result, driver, cardTerminalType, timeoutSeconds)
                } else {
                    result.error("INVALID_DEVICE", "Invalid driver, cardTerminalType or timeoutSeconds parameter", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }   

    private fun connectToDevice(
        result: Result, 
        driverMap: Map<String, Any>,
        cardTerminalType: Int, 
        timeoutSeconds: Int,
        ) {
        val card = driverMap["card"] as? String
        val name = driverMap["name"] as? String
        val firstName = driverMap["firstName"] as? String
        val email = driverMap["email"] as? String
        val phone = driverMap["phone"] as? String
        val agencyID = driverMap["agencyID"] as? String

        if (card != null && firstName != null && name != null && email != null && phone != null && agencyID != null) {
            val driver = Driver(
                card = card,
                name = name,
                firstName = firstName,
                email = email,
                phone = phone,
                agencyID = agencyID
            )
            smartCardReader.connectToDevice(activity, context, driver, cardTerminalType, timeoutSeconds)
        } else {
            result.error("INVALID_DEVICE", "Invalid parameters $agencyID", null)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
